unittests/linuxcrosstests.py

```diff
@@ -116,7 +117,11 @@ def test_run_native_test(self):
 
 
 def should_run_cross_mingw_tests():
-    return shutil.which('x86_64-w64-mingw32-gcc') and not (is_windows() or is_cygwin())
+    # The compiler binary could be a symlink to ccache, in which case
+    # it will appear to exist but will fail when called.
+    return shutil.which('x86_64-w64-mingw32-gcc') and \
+        subprocess.call(['x86_64-w64-mingw32-gcc', '--version'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0 and \
+        not (is_windows() or is_cygwin())
```

Comment:
Reading the docs I think `os.path.exists(shutil.which(…))` should be sufficient. It’s supposed to resolve symlinks correctly

Reply:
Not if it's a symlink to ccache though.

That being said I think just running it should be enough. No need to also use shutil.which?

Reply:
> That being said I think just running it should be enough. No need to also use shutil.which?

done that way

Reply:
Pretty sure you do have to catch FileNotFoundError, which doesn't involve stat'ing the filesystem to see if it exists like shutil.which does, but is still some code to include.

Reply:
(This is probably a good thing since the logic was getting quite hairy and nested. :P)

Reply:
When I try this in isolation I do not get an exception:

```python
import subprocess

ret = subprocess.call(['x86_64-w64-mingw32-gcc', '--version'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
print(ret)
```

just prints `1` without an exception.

Reply:
Sorry, that didn't make any sense. I have pushed a new version with the exception check.

Personally, I don't think that turned out to be better than the version that checks `shutil.which` followed by `subprocess.call`, but I'm happy to oblige.

I also noticed that the `is_windows()` check in the `skipUnless` right below is redundant.

Finally, there is a similar issue in `should_run_cross_arm_tests` in the same file. It doesn't affect my system, but it would probably make sense to keep all these "is the compiler there" checks similar?

Reply:
> Personally, I don't think that turned out to be better than the version that checks `shutil.which` followed by `subprocess.call`, but I'm happy to oblige.

Hmm, that version had chained `and`s running across three lines, and a comment that was distanced from the line it was commenting on. Interesting that you prefer that.

Reply:
> Finally, there is a similar issue in `should_run_cross_arm_tests` in the same file. It doesn't affect my system, but it would probably make sense to keep all these "is the compiler there" checks similar?

I have also changed that function in the updated version of the patch. This is complete from my perspective now.